File: python/house/house.py
# Define the possible actions that each item in the nursery rhyme can perform.
ACTIONS = [
    "lay in",
    "ate",
    "killed",
    "worried",
    "tossed",
    "milked",
    "kissed",
    "married",
    "woke",
    "kept",
    "belonged to",
    "",
]

# Define the items that make up the nursery rhyme.
THINGS = [
    "house that Jack built.",
    "malt",
    "rat",
    "cat",
    "dog",
    "cow with the crumpled horn",
    "maiden all forlorn",
    "man all tattered and torn",
    "priest all shaven and shorn",
    "rooster that crowed in the morn",
    "farmer sowing his corn",
    "horse and the hound and the horn",
]

# Function to recite a range of verses from the nursery rhyme.
def recite(start_verse, end_verse):
    return [verse(n) for n in range(start_verse - 1, end_verse)]

# Function to generate a single verse of the nursery rhyme.
def verse(n):
    return ' '.join(
        ["This is the {}".format(THINGS[n])] +
        ["that {} the {}".format(ACTIONS[i], THINGS[i])
         for i in range(n)][::-1]
    )

File: javascript/house/house.js
const RHYME_DATA = [
  // This array stores the rhyme data, including the character and the action associated with each line
  {
    character: "house that Jack built.",
    action: null // Null indicates the starting line
  },
  {
    character: "malt",
    action: "lay in" // Action describes the relationship between the current and previous line
  },
  {
    character: "rat",
    action: "ate"
  },
  {
    character: "cat",
    action: "killed"
  },
  {
    character: "dog",
    action: "worried"
  },
  {
    character: "cow with the crumpled horn",
    action: "tossed"
  },
  {
    character: "maiden all forlorn",
    action: "milked"
  },
  {
    character: "man all tattered and torn",
    action: "kissed"
  },
  {
    character: "priest all shaven and shorn",
    action: "married"
  },
  {
    character: "rooster that crowed in the morn",
    action: "woke"
  },
  {
    character: "farmer sowing his corn",
    action: "kept"
  },
  {
    character: "horse and the hound and the horn",
    action: "belonged to"
  }
];

export class House {
  // This class encapsulates the functionality for generating the rhyme verses

  static verse(num) {
    // This method generates a single verse of the rhyme, given the verse number
    if (num < 1) return []; // Return an empty array if the verse number is less than 1

    // Construct the first line of the verse
    const firstLine = `This is the ${RHYME_DATA[num - 1].character}`;

    // Construct the remaining lines using the _remainingLines method
    const remainingLines = this._remainingLines(num - 1);

    // Combine the first line and the remaining lines to form the verse
    const verseLines = [firstLine, ...remainingLines];

    return verseLines;
  }

  static _remainingLines(num) {
    // This recursive method generates the remaining lines of the verse, given the verse number
    if (num < 1) return []; // Return an empty array if the verse number is less than 1

    // Construct the first remaining line
    const firstRemainingLine = `that ${RHYME_DATA[num].action} the ${RHYME_DATA[num - 1].character}`;

    // Construct the remaining lines using the _remainingLines method
    const remainingLines = this._remainingLines(num - 1);

    // Combine the first remaining line and the remaining lines to form the remaining verse lines
    const verseLines = [firstRemainingLine, ...remainingLines];

    return verseLines;
  }

  static verses(start, end) {
    // This method generates a sequence of verses, given the start and end verse numbers
    let result = [];

    for (let i = start; i <= end; i++) {
      // Generate each verse and append it to the result array
      result.push(...this.verse(i), "");
    }

    // Remove the last empty string added to the result array
    result.pop();

    return result; // Return the sequence of verses
  }
}
